components/brave_shields/adblock/rs/src/engine.rs

```diff
@@ -0,0 +1,195 @@
+use std::str::Utf8Error;
+
+use adblock::engine::Engine as InnerEngine;
+use adblock::lists::FilterSet;
+use adblock::resources::{MimeType, Resource, ResourceType};
+use adblock::url_parser::ResolvesDomain;
+use cxx::{let_cxx_string, CxxString, CxxVector};
+
+use crate::ffi::{
+    resolve_domain_position, BlockerDebugInfo, BlockerResult, BoxEngineResult, EmptyTupleResult,
+    FilterListMetadata, RegexManagerDiscardPolicy, VecStringResult,
+};
+use crate::result::InternalError;
+
+pub struct Engine {
+    engine: InnerEngine,
+}
+
+impl Default for Box<Engine> {
+    fn default() -> Self {
+        new_engine()
+    }
+}
+
+pub fn new_engine() -> Box<Engine> {
+    Box::new(Engine { engine: InnerEngine::new(true) })
+}
+
+pub fn engine_with_rules(rules: &CxxVector<u8>) -> BoxEngineResult {
+    || -> Result<Box<Engine>, InternalError> {
+        let mut filter_set = FilterSet::new(false);
+        filter_set.add_filter_list(std::str::from_utf8(rules.as_slice())?, Default::default());
+        let engine = InnerEngine::from_filter_set(filter_set, true);
+        Ok(Box::new(Engine { engine }))
+    }()
+    .into()
+}
+
+struct DomainResolver;
+
+impl ResolvesDomain for DomainResolver {
+    fn get_host_domain(&self, host: &str) -> (usize, usize) {
+        let_cxx_string!(host_cxx_string = host);
+        let position = resolve_domain_position(&host_cxx_string);
+        (position.start as usize, position.end as usize)
+    }
+}
+
+pub fn set_domain_resolver() -> bool {
+    adblock::url_parser::set_domain_resolver(Box::new(DomainResolver)).is_ok()
+}
+
+pub fn read_list_metadata(list: &CxxVector<u8>) -> FilterListMetadata {
+    std::str::from_utf8(list.as_slice())
+        .map(|list| adblock::lists::read_list_metadata(list).into())
+        .unwrap_or_default()
+}
+
+fn convert_cxx_string_vector_to_string_collection<C>(
+    value: &CxxVector<CxxString>,
+) -> Result<C, Utf8Error>
+where
+    C: FromIterator<String>,
+{
+    value.iter().map(|s| s.to_str().map(|t| t.to_string())).collect()
+}
+
+impl Engine {
+    pub fn enable_tag(&mut self, tag: &CxxString) {
+        self.engine.enable_tags(&[tag.to_str().unwrap()])
+    }
+
+    pub fn disable_tag(&mut self, tag: &CxxString) {
+        self.engine.disable_tags(&[tag.to_str().unwrap()])
+    }
+
+    pub fn tag_exists(&self, key: &CxxString) -> bool {
+        self.engine.tag_exists(key.to_str().unwrap())
+    }
+
+    pub fn matches(
+        &self,
+        url: &CxxString,
+        hostname: &CxxString,
+        source_hostname: &CxxString,
+        request_type: &CxxString,
+        third_party_request: bool,
+        previously_matched_rule: bool,
+        force_check_exceptions: bool,
+    ) -> BlockerResult {
+        self.engine
+            .check_network_urls_with_hostnames_subset(
+                url.to_str().unwrap(),
+                hostname.to_str().unwrap(),
+                source_hostname.to_str().unwrap(),
+                request_type.to_str().unwrap(),
```

Comment:
Looks like these come from a GURL, which should be ascii, but the unwraps make me nervous. This is a public function and it would be easy to start passing it user-defined data and take down the browser process.

If you're sure this is safe, please add a comment explaining how and what the constraints are, so if it comes up in an audit or crash it's clear what the expectation is.

I'd feel better if these were `unwrap_or("")` with some kind of debug-only assert if you want to detect violations in testing

Reply:
duplicate of https://github.com/brave/brave-core/pull/17368#discussion_r1157694477